README.md
thecity-login-wordpress
=======================

A WordPress plugin that allows users to login direclty to The City from your WordPress website.

changelog
0.1
* Initial release of the TheCity-Login-Wordpress plugin.

0.2
* Fix issue if widget was not on page "Subdomain not set" was displayed.

0.3
* Merged in @markjflores design additions to use The City styling.
* Refactored login class to extract out designs.

widget_info.php
<?php
  $subdomain_key = trim($subdomain_key);
  $login_display_choice = trim($login_display_choice);
  $show_remember_me = $show_remember_me == 'on' ? '1' : '0';
  $use_placeholder_text = $use_placeholder_text == 'on' ? '1' : '0';

  $form_display = '';
  switch($login_display_choice) {
    case "plain":
      $form_display = 'thecity_login_normal_plain';
      break;

    case "inline":
      $form_display = 'thecity_login_inline_plain';
      break;

    case "city_style_normal":
      $form_display = 'city_style_normal';
      break;  

    case "city_style_inline":
      $form_display = 'city_style_inline';
      break;        

    default: // plane
      $form_display = 'thecity_login_normal_plain';
  }

  if( empty($subdomain_key)) { echo 'City subdomain not set'; } else {  ?>
  <div id="thecity_login" 
       class="<?php echo $form_display; ?>" 
       subdomain="<?php echo $subdomain_key; ?>" 
       show_remember_me="<?php echo $show_remember_me; ?>" 
       use_placeholder_text="<?php echo $use_placeholder_text; ?>" ></div>
<?php } ?>

thecity_login.php
<?php
/*
Plugin Name: The City Login Widget
Plugin URI: http://developer.onthecity.org/thecity-plugins/wordpress-login/
Description: A WordPress plugin that allows users to login directly to The City from your WordPress website.
Author: Wes Hays
Version: 0.3
Author URI: http://www.OnTheCity.org
*/

include_once 'thecity_login_scripts.php';


class The_City_Login_Widget extends WP_Widget {
  
  function __construct() {
    $widget_ops = array('classname' => 'the_city_login_widget', 
                        'description' => 'A WordPress plugin that allows users to login directly to The City from your WordPress website.' );
    $this->WP_Widget('the-city-login-widget', 'The City Login', $widget_ops);
  }
  

  function form($instance) {
    /* Set up some default widget settings. */
		$defaults = array( 'subdomain_key' => '', 
                       'login_display_choice' => '');

		$instance = wp_parse_args( (array) $instance, $defaults );    

    $subdomain_key = strip_tags($instance['subdomain_key']);
    $login_display_choice = strip_tags($instance['login_display_choice']);
    $show_remember_me = strip_tags($instance['show_remember_me']);
    $use_placeholder_text = strip_tags($instance['use_placeholder_text']);
    ?>

   <p>
     <label for="<?php echo $this->get_field_id('subdomain_key'); ?>">
       Subdomain: 
       <input class="widefat" 
              id="<?php echo $this->get_field_id('subdomain_key'); ?>" 
              name="<?php echo $this->get_field_name('subdomain_key'); ?>" 
              type="text" 
              value="<?php echo attribute_escape($subdomain_key); ?>" />
      </label>
      <i>Ex: https://[subdomain].OnTheCity.org</i>
    </p>


    <?php 
      $plain_s = $inline_s = '';
      switch($instance['login_display_choice']) {
        case 'plain':
          $plain_s = 'selected="selected"'; 
          break;
        case 'inline':
          $inline_s = 'selected="selected"'; 
          break;
        case 'city_style_normal':
          $city_style_normal = 'selected="selected"'; 
          break;
        case 'city_style_inline':
          $city_style_inline = 'selected="selected"'; 
          break;
      }
    ?> 

    <p>    
      <label for="<?php echo $this->get_field_id('login_display_choice'); ?>">
        Display:        			
        <select class="widefat" 
                id="<?php echo $this->get_field_id('login_display_choice'); ?>" 
                name="<?php echo $this->get_field_name('login_display_choice'); ?>">
            <option value="plain" <?php echo $plain_s; ?> >Plain</option>
        		<option value="inline" <?php echo $inline_s; ?> >Inline</option>
            <option value="city_style_normal" <?php echo $city_style_normal; ?> >City Style Normal</option>
            <option value="city_style_inline" <?php echo $city_style_inline; ?> >City Style Inline</option>
        </select>
      </label>    
    </p>    

    <?php $show_remember_me_checked = empty($show_remember_me) ? '' : 'checked="checked"'; ?>
    <p>
      <label for="<?php echo $this->get_field_id('show_remember_me'); ?>">
        <input type="checkbox" 
               id="<?php echo $this->get_field_id('show_remember_me'); ?>" 
               name="<?php echo $this->get_field_name('show_remember_me'); ?>"
               <?php echo $show_remember_me_checked ?> /> Show Remember me checkbox       
      </label>
    <p>

    <?php $use_placeholder_text_checked = empty($use_placeholder_text) ? '' : 'checked="checked"'; ?>
    <p>
      <label for="<?php echo $this->get_field_id('use_placeholder_text'); ?>">
        <input type="checkbox" 
               id="<?php echo $this->get_field_id('use_placeholder_text'); ?>" 
               name="<?php echo $this->get_field_name('use_placeholder_text'); ?>"
               <?php echo $use_placeholder_text_checked ?> /> Use placeholder text       
      </label>
    <p>      

    <?php
  }
  

  function update($new_instance, $old_instance) {
    $instance = $old_instance;
    $instance['subdomain_key'] = strip_tags($new_instance['subdomain_key']);
    $instance['login_display_choice'] = strip_tags($new_instance['login_display_choice']);
    $instance['show_remember_me'] = strip_tags($new_instance['show_remember_me']);
    $instance['use_placeholder_text'] = strip_tags($new_instance['use_placeholder_text']);
    return $instance;
  }
  


  function widget($args, $instance) {
    extract($args);
    $subdomain_key = empty($instance['subdomain_key']) ? ' ' : $instance['subdomain_key'];
    $login_display_choice = empty($instance['login_display_choice']) ? 'plain' : $instance['login_display_choice'];
    $show_remember_me = empty($instance['show_remember_me']) ? ' ' : $instance['show_remember_me'];
    $use_placeholder_text = empty($instance['use_placeholder_text']) ? ' ' : $instance['use_placeholder_text'];
    include dirname(__FILE__).'/widget_info.php';
    echo $after_widget;
  }
  
}

add_action('widgets_init', create_function('', 'return register_widget("The_City_Login_Widget");'));

?>

scripts/tc_base64.js
var TCBase64 = {
 
  // private property
  _keyStr : "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",
 
  // public method for encoding
  encode : function (input) {
    var output = "";
    var chr1, chr2, chr3, enc1, enc2, enc3, enc4;
    var i = 0;
 
    input = TCBase64._utf8_encode(input);
 
    while (i < input.length) {
 
      chr1 = input.charCodeAt(i++);
      chr2 = input.charCodeAt(i++);
      chr3 = input.charCodeAt(i++);
 
      enc1 = chr1 >> 2;
      enc2 = ((chr1 & 3) << 4) | (chr2 >> 4);
      enc3 = ((chr2 & 15) << 2) | (chr3 >> 6);
      enc4 = chr3 & 63;
 
      if (isNaN(chr2)) {
        enc3 = enc4 = 64;
      } else if (isNaN(chr3)) {
        enc4 = 64;
      }
 
      output = output +
      this._keyStr.charAt(enc1) + this._keyStr.charAt(enc2) +
      this._keyStr.charAt(enc3) + this._keyStr.charAt(enc4);
 
    }
 
    return output;
  },
 
  // public method for decoding
  decode : function (input) {
    var output = "";
    var chr1, chr2, chr3;
    var enc1, enc2, enc3, enc4;
    var i = 0;
 
    input = input.replace(/[^A-Za-z0-9\+\/\=]/g, "");
 
    while (i < input.length) {
 
      enc1 = this._keyStr.indexOf(input.charAt(i++));
      enc2 = this._keyStr.indexOf(input.charAt(i++));
      enc3 = this._keyStr.indexOf(input.charAt(i++));
      enc4 = this._keyStr.indexOf(input.charAt(i++));
 
      chr1 = (enc1 << 2) | (enc2 >> 4);
      chr2 = ((enc2 & 15) << 4) | (enc3 >> 2);
      chr3 = ((enc3 & 3) << 6) | enc4;
 
      output = output + String.fromCharCode(chr1);
 
      if (enc3 != 64) {
        output = output + String.fromCharCode(chr2);
      }
      if (enc4 != 64) {
        output = output + String.fromCharCode(chr3);
      }
 
    }
 
    output = TCBase64._utf8_decode(output);
 
    return output;
 
  },
 
  // private method for UTF-8 encoding
  _utf8_encode : function (string) {
    string = string.replace(/\r\n/g,"\n");
    var utftext = "";
 
    for (var n = 0; n < string.length; n++) {
 
      var c = string.charCodeAt(n);
 
      if (c < 128) {
        utftext += String.fromCharCode(c);
      }
      else if((c > 127) && (c < 2048)) {
        utftext += String.fromCharCode((c >> 6) | 192);
        utftext += String.fromCharCode((c & 63) | 128);
      }
      else {
        utftext += String.fromCharCode((c >> 12) | 224);
        utftext += String.fromCharCode(((c >> 6) & 63) | 128);
        utftext += String.fromCharCode((c & 63) | 128);
      }
 
    }
 
    return utftext;
  },
 
  // private method for UTF-8 decoding
  _utf8_decode : function (utftext) {
    var string = "";
    var i = 0;
    var c = c1 = c2 = 0;
 
    while ( i < utftext.length ) {
 
      c = utftext.charCodeAt(i);
 
      if (c < 128) {
        string += String.fromCharCode(c);
        i++;
      }
      else if((c > 191) && (c < 224)) {
        c2 = utftext.charCodeAt(i+1);
        string += String.fromCharCode(((c & 31) << 6) | (c2 & 63));
        i += 2;
      }
      else {
        c2 = utftext.charCodeAt(i+1);
        c3 = utftext.charCodeAt(i+2);
        string += String.fromCharCode(((c & 15) << 12) | ((c2 & 63) << 6) | (c3 & 63));
        i += 3;
      }
 
    }
 
    return string;
  }
 
}

thecity_login_scripts.php
<?php

function thecity_login_wordpress_scripts() {     
  wp_register_script('thecity_login_js', plugins_url('/the-city-login/scripts/thecity_login.js'));   
  wp_register_script('thecity_login_load_js', plugins_url('/the-city-login/scripts/thecity_login_load.js')); 
  wp_register_script('formatter_city_style_js', plugins_url('/the-city-login/scripts/formatter_city_style.js')); 

  wp_enqueue_script( array('jquery') );  
  wp_enqueue_script('thecity_login_js');    
  wp_enqueue_script('thecity_login_load_js');  
  wp_enqueue_script('formatter_city_style_js');  
}

add_action('wp_enqueue_scripts', 'thecity_login_wordpress_scripts');  



function thecity_login_wordpress_styles() {  
  wp_register_style('thecity_login_wordpress_style', plugins_url( '/the-city-login/scripts/thecity_login.css'));   
  wp_register_style('thecity_appkit', plugins_url( '/the-city-login/scripts/thecity_appkit.css'));     
  wp_register_style('formatter_city_style_css', plugins_url( '/the-city-login/scripts/formatter_city_style.css'));   

  wp_enqueue_style('thecity_login_wordpress_style'); 
  wp_enqueue_style('thecity_appkit');  
  wp_enqueue_style('formatter_city_style_css');  
}

add_action('wp_enqueue_scripts', 'thecity_login_wordpress_styles');  

?>

scripts/formatter_city_style.js
function formatter_city_style_normal(self) { _city_style_render(self); }
function formatter_city_style_inline(self) { _city_style_render(self); }

function _city_style_render(self) {
  console.log("DISPLAY: " + self.defaults['display_style'] );
  if(self.defaults['display_style'] == 'city_style_inline') {
    jQuery('#'+self.defaults['city_login_div_id']+' li:first').addClass('grid-6-12');
    jQuery('#'+self.defaults['city_login_div_id']+' li:nth-child(2)').addClass('grid-6-12');
  } else { // city_style_normal
    jQuery('#'+self.defaults['city_login_div_id']+' li:first').addClass('grid-12-12');
    jQuery('#'+self.defaults['city_login_div_id']+' li:nth-child(2)').addClass('grid-12-12');
  }

  // Universal styling
  jQuery('#'+self.defaults['city_login_div_id']+' form').css({"padding": "10px", "background-color": "#FFFFFF"});
  jQuery('#'+self.defaults['city_login_div_id']+' li:nth-child(3)').addClass('formee-list');
  jQuery('#'+self.defaults['city_login_div_id']+' li:last').addClass('form-footer');
  jQuery('#'+self.defaults['city_login_div_id']+' #city_login_link').addClass('btn').addClass('btn_go').addClass('right');    
}

scripts/thecity_login.js
// **************************************************
// TheCity Login JavaScript Wrapper
// **************************************************
TheCityLogin = {
  defaults : { 'subdomain' : null,
               'city_login_div_id' : 'thecity_login',
               'display_form_loading_message' : true,
               'show_remember_me' : true,
               'use_placeholder_text' : false},
               
  _class_vars : {'login_form_url' : 'http://authentication.onthecity.org/sessions/remote_form.json?callback=?',
                 'login_url' : 'http://authentication.onthecity.org/sessions/remote_login.json?callback=?'}, 
  
  start : function(options) {
    if(options !== undefined) { 
      jQuery.extend(this.defaults, options)  
    }
    
    if(this.defaults['subdomain'] == null) {
      alert('Subdomain must be set');
      return;
    }

    if(this.defaults['city_login_div_id'] == null || this.defaults['city_login_div_id'] == '') {
      alert('Div ID not specified');
      return;
    }    

    if(this.defaults['display_form_loading_message']) {
      jQuery('#'+this.defaults['city_login_div_id']).html("Loading login form....");
    }

    var self = this; // Need object scope at this point
    jQuery.getJSON(this._class_vars['login_form_url'], function(response){        
        var form = response['form'];
        form = form.replace(/[\r\n]/g, '');
        form = TCBase64.decode(form);
        jQuery('#'+self.defaults['city_login_div_id']).html(form);

        if(!self.defaults['show_remember_me']) {
          jQuery('#'+self.defaults['city_login_div_id']+' #authcity #remember-me').hide();
        }

        if(self.defaults['use_placeholder_text']) {
          jQuery('#'+self.defaults['city_login_div_id']+' label[for="login"]').hide();
          jQuery('#'+self.defaults['city_login_div_id']+' label[for="password"]').hide();
          jQuery('#'+self.defaults['city_login_div_id']+' #authcity #login').attr('placeholder','Login');
          jQuery('#'+self.defaults['city_login_div_id']+' #authcity #password').attr('placeholder','Password');
        }
        
        self._add_meta_tag_csrf_token();
        self._add_login_link_listener();
        self._apply_styling_formatter();
      }
    )   
  },

  _add_meta_tag_csrf_token : function() {
    var csrf_token = document.createElement('meta');
    csrf_token.name = "csrf-token";
    csrf_token.content = jQuery('[name="authenticity_token"]').val()
    document.getElementsByTagName('head')[0].appendChild(csrf_token);
  },  

  _remove_errors : function() {
    jQuery(".city_error").remove();
  },  

  _add_error : function(msg) {
    var self = this; // Need object scope at this point
    var error_li = document.createElement("li");
    error_li.setAttribute("class","city_error");
    error_li.innerHTML = msg;
    jQuery("#authcity ul ").prepend( error_li );
  },  

  _add_login_link_listener : function() {
    var self = this; // Need object scope at this point
    jQuery("#city_login_link").click(function() {
      self._remove_errors();

      var params = {
        'login' : jQuery("#login").val(),
        'password' : jQuery("#password").val(),
        'authenticity_token' : jQuery('[name="authenticity_token"]').val()
      };

      jQuery.getJSON(self._class_vars['login_url'], params, function(response){   
        if(response["success"] != undefined) {
          window.location = "https://" + self.defaults['subdomain'] + ".onthecity.org";
        } else {
          self._add_error(response["errors"]);
        }          
      });      
    });
  },

  _apply_styling_formatter : function() {
    var func = 'formatter_' + this.defaults['display_style'];
    if (typeof(window[func]) === 'function') { window[func](this); }
  }
  
}




// **************************************************
// Base64 Utility Section
// **************************************************
var TCBase64 = {
 
  // private property
  _keyStr : "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",
 
  // public method for encoding
  encode : function (input) {
    var output = "";
    var chr1, chr2, chr3, enc1, enc2, enc3, enc4;
    var i = 0;
 
    input = TCBase64._utf8_encode(input);
 
    while (i < input.length) {
 
      chr1 = input.charCodeAt(i++);
      chr2 = input.charCodeAt(i++);
      chr3 = input.charCodeAt(i++);
 
      enc1 = chr1 >> 2;
      enc2 = ((chr1 & 3) << 4) | (chr2 >> 4);
      enc3 = ((chr2 & 15) << 2) | (chr3 >> 6);
      enc4 = chr3 & 63;
 
      if (isNaN(chr2)) {
        enc3 = enc4 = 64;
      } else if (isNaN(chr3)) {
        enc4 = 64;
      }
 
      output = output +
      this._keyStr.charAt(enc1) + this._keyStr.charAt(enc2) +
      this._keyStr.charAt(enc3) + this._keyStr.charAt(enc4);
 
    }
 
    return output;
  },
 
  // public method for decoding
  decode : function (input) {
    var output = "";
    var chr1, chr2, chr3;
    var enc1, enc2, enc3, enc4;
    var i = 0;
 
    input = input.replace(/[^A-Za-z0-9\+\/\=]/g, "");
 
    while (i < input.length) {
 
      enc1 = this._keyStr.indexOf(input.charAt(i++));
      enc2 = this._keyStr.indexOf(input.charAt(i++));
      enc3 = this._keyStr.indexOf(input.charAt(i++));
      enc4 = this._keyStr.indexOf(input.charAt(i++));
 
      chr1 = (enc1 << 2) | (enc2 >> 4);
      chr2 = ((enc2 & 15) << 4) | (enc3 >> 2);
      chr3 = ((enc3 & 3) << 6) | enc4;
 
      output = output + String.fromCharCode(chr1);
 
      if (enc3 != 64) {
        output = output + String.fromCharCode(chr2);
      }
      if (enc4 != 64) {
        output = output + String.fromCharCode(chr3);
      }
 
    }
 
    output = TCBase64._utf8_decode(output);
 
    return output;
 
  },
 
  // private method for UTF-8 encoding
  _utf8_encode : function (string) {
    string = string.replace(/\r\n/g,"\n");
    var utftext = "";
 
    for (var n = 0; n < string.length; n++) {
 
      var c = string.charCodeAt(n);
 
      if (c < 128) {
        utftext += String.fromCharCode(c);
      }
      else if((c > 127) && (c < 2048)) {
        utftext += String.fromCharCode((c >> 6) | 192);
        utftext += String.fromCharCode((c & 63) | 128);
      }
      else {
        utftext += String.fromCharCode((c >> 12) | 224);
        utftext += String.fromCharCode(((c >> 6) & 63) | 128);
        utftext += String.fromCharCode((c & 63) | 128);
      }
 
    }
 
    return utftext;
  },
 
  // private method for UTF-8 decoding
  _utf8_decode : function (utftext) {
    var string = "";
    var i = 0;
    var c = c1 = c2 = 0;
 
    while ( i < utftext.length ) {
 
      c = utftext.charCodeAt(i);
 
      if (c < 128) {
        string += String.fromCharCode(c);
        i++;
      }
      else if((c > 191) && (c < 224)) {
        c2 = utftext.charCodeAt(i+1);
        string += String.fromCharCode(((c & 31) << 6) | (c2 & 63));
        i += 2;
      }
      else {
        c2 = utftext.charCodeAt(i+1);
        c3 = utftext.charCodeAt(i+2);
        string += String.fromCharCode(((c & 15) << 12) | ((c2 & 63) << 6) | (c3 & 63));
        i += 3;
      }
 
    }
 
    return string;
  }
 
}
